src/main/java/org/ambraproject/wombat/service/CommentServiceImpl.java

```diff
@@ -128,6 +145,28 @@ private void addCreatorData(Collection<Map<String, Object>> rootComments) {
     }));
   }
 
+  private void addArticleTitle(Collection<Map<String, Object>> rootComments) {
+    // Gather up all distinct article DOIs in the forest
+    Set<String> articleDois = Collections.synchronizedSet(new HashSet<>());
```

Comment:
You may get better performance by using `java.util.concurrent.ConcurrentHashMap`, rather than the **synchronized** version, as the **concurrent** version will avoid locking the entire set.
```
Set<String> articleDois = ConcurrentHashMap.newKeySet();
```

Reply:
changed.